### PR TITLE
[Feature][SFA] Support DCP with replicate-indexer for SFA

### PR DESCRIPTION
**Pull request by pisceskkk**
### What this PR does / why we need it?

This PR adds SFA DCP replicate-indexer support for SFA models.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Validated with the GLM-5:

- model: GLM-5 W4A8
- parallelism: `-tp 16 -pcp 1 -dcp 16`
- key options: `enable_dsa_cp`, `enable_sparse_c8`, FULL_DECODE_ONLY graph mode, 3-token MTP speculative decoding

GSM8K:
| dataset | version | metric | mode | vllm-api-general-chat |
|----- | ----- | ----- | ----- | -----|
| gsm8k | 50861a | accuracy | gen | 93.40 |

GPQA
| dataset | version | metric | mode | vllm-api-general-chat |
|----- | ----- | ----- | ----- | -----|
| GPQA_diamond | b1ed2c | accuracy | gen | 85.86 |

Benchmark for GLM-5.2 W4A8 with 1M context (with https://github.com/vllm-project/vllm-ascend/pull/11496, which is performance optimization in llm_base_proposer for general CP logic).
Notes: TOPT will increase to ~29ms without #11496.
```
export VLLM_ASCEND_ENABLE_FLASHCOMM1=0
export VLLM_ASCEND_ENABLE_FUSED_MC2=0
vllm serve /mnt/share/weight/GLM-5.2-Provide-0610-W4A8 \
    --seed 1024 \
    --host 0.0.0.0 \
    --port 9000 \
    --served-model-name glm-5 \
    --max_model_len 1024000 \
    --max-num-batched-tokens 2048 \
    --gpu-memory-utilization 0.90 \
    --api-server-count 1 \
    --max-num-seqs 4 \
    -dp 1 -pp 1 -tp 16 -pcp 1 -dcp 16 \
    --cp-kv-cache-interleave-size 128 \
    --compilation-config '{"cudagraph_mode": "FULL_DECODE_ONLY", "cudagraph_capture_sizes":[4, 16]}' \
    --additional-config '{"enable_dsa_cp": false, "ascend_compilation_config":{"enable_npugraph_ex": true, "enable_static_kernel": false},"fuse_muls_add":true, "multistream_overlap_shared_expert": true, "enable_mc2_hierarchy_comm":false, "enable_sparse_c8": true, "enable_cpu_binding":true, "recompute_scheduler_enable": false}' \
    --speculative-config '{"num_speculative_tokens": 3,"method": "deepseek_mtp"}' \
    --quantization ascend \
    --enable-expert-parallel \
    --safetensors-load-strategy 'prefetch' \
    2>&1 | tee server.log

```

| case | req num | TTFT avg/P90 ms | TPOT avg/P90 ms | output tps | E2E s |
|---|---:|---:|---:|---:|---:|
| 16k/1k | 5 | 7822.0 / 8007.2 | 22.9 / 23.5 | 32.5895 | 153.42 |
| 64k/1k | 5 | 38108.7 / 38240.5 | 24.0 / 25.4 | 16.1141 | 310.29 |
| 128k/1k | 5 | 83483.3 / 83703.2 | 24.2 / 26.4 | 9.2843 | 538.54 |
| 200k/1k | 5 | 140338.6 / 140444.2 | 23.9 / 25.0 | 6.0901 | 821.01 |
| 1M/1k | 1 | 1201835.6 / 1201835.6 | 23.8 / 23.8 | 0.8159 | 1225.64 |

- vLLM version: v0.23.0
- vLLM main: https://github.com/vllm-project/vllm/commit/1f486d96a17303ce8db8e02be39545b2be338446

### CODE REVIEW

File: vllm_ascend/attention/sfa_v1.py [hunk 1]
```diff
@@ -1621,9 +1648,27 @@ def forward(
...
+        else:
+            attn_output = self._execute_sparse_flash_attention_process(
+                ql_nope,
+                q_pe,
+                kv_cache,
+                topk_indices,
+                attn_metadata,
+                actual_seq_lengths_query,
+                actual_seq_lengths_key,
+                query_gather_context=dcp_query_gather_context,
```
Comment (pisceskkk) [inline vllm_ascend/attention/sfa_v1.py]: move `query_gather_context` into `dcp_context` and handle it in `sfa_cp.py`

**Comment by linfeng-yuan** [comment]
1. Remove redundant code in mooncake
2. Split llm_base_proposer optimization in a new PR
3. Check AscendC Modification with @maoxx241 
4. SFA modification review @dragondream-chen 
5. spec decoding @lilinsiman 
6. CP @weiguihua2

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 2]
```diff
@@ -578,3 +593,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+            0,
+            local_block_table_cols,
+        )
+
+        total_cp_size = self.pcp_size * self.dcp_size
+        blocks_per_phys_block = self.blocks_per_phys_block
+        block_table_cols = block_table_replicated_view.shape[1]
+        replicated_col_idx = torch.arange(
+            block_table_cols,
+            dtype=torch.long,
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]: Do we need `torch.long` as dtype for `replicated_col_idx`? will it exceed max_int32?

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 3]
```diff
@@ -578,3 +593,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+        block_dtype = local_logical_blocks.dtype
+        rank_in_replicated_view = rank_in_replicated_view.to(block_dtype)
+        if blocks_per_phys_block == 1:
+            replicated_blocks = local_logical_blocks * total_cp_size + rank_in_replicated_view
+        else:
+            local_sub_blocks = local_logical_blocks % blocks_per_phys_block
+            local_phys_blocks = local_logical_blocks // blocks_per_phys_block
+            replicated_blocks = (
+                local_phys_blocks * total_cp_size + rank_in_replicated_view
+            ) * blocks_per_phys_block + local_sub_blocks
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py:743]: I think the logic could be unified as the branch else

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 4]
```diff
@@ -578,3 +593,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+        slot_mapping_replicated_view.fill_(-1)
+        if num_actual_tokens == 0:
+            return slot_mapping_replicated_view
+
+        query_lens = (
+            common_attn_metadata.query_start_loc[1 : num_reqs + 1] - common_attn_metadata.query_start_loc[:num_reqs]
+        )
+        req_indices = torch.repeat_interleave(
+            torch.arange(num_reqs, dtype=torch.long, device=self.device),
+            query_lens.to(device=self.device, dtype=torch.long),
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]: ditto.

File: vllm_ascend/attention/context_parallel/sfa_cp.py (hunk 3 above)
Comment (pisceskkk) [reply vllm_ascend/attention/context_parallel/sfa_cp.py:743]: In most cases, blocks_per_phys_block == 1. And this branch can reduce more integer calculation, which has a significant impact on performance. So I think it may be better to keep this branch.

File: vllm_ascend/attention/context_parallel/sfa_cp.py (hunk 2 above)
Comment (pisceskkk) [reply vllm_ascend/attention/context_parallel/sfa_cp.py]: yes, torch.int32 is big enough, have changed.

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 5]
```diff
@@ -578,3 +594,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+        )
+        local_col_idx = (
+            replicated_col_idx // (total_cp_size * blocks_per_phys_block) * blocks_per_phys_block
+            + replicated_col_idx % blocks_per_phys_block
+        )
+        rank_in_replicated_view = (replicated_col_idx // blocks_per_phys_block) % total_cp_size
+
+        local_logical_blocks = torch.index_select(dcp_block_table, 1, local_col_idx)
+        block_dtype = local_logical_blocks.dtype
+        rank_in_replicated_view = rank_in_replicated_view.to(block_dtype)
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]: no need for this dtype cast now.

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 6]
```diff
@@ -578,3 +594,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+            device=self.device,
+            dtype=torch.int32,
+        )
+        logical_block_idx = positions // self.replicated_view_block_size
+        block_offsets = positions % self.replicated_view_block_size
+        block_table_indices = req_indices * block_table_replicated_view.shape[1] + logical_block_idx
+        block_numbers = block_table_replicated_view.flatten()[block_table_indices]
+        slot_mapping_replicated_view[:num_actual_tokens] = (
+            block_numbers * self.replicated_view_block_size + block_offsets
+        ).to(slot_mapping_replicated_view.dtype)
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]: no need for this dtype cast.

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 7]
```diff
@@ -578,3 +594,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+        key = (str(device), dtype)
+        constants = self._dcp_remap_scalar_tensors.get(key)
+        if constants is not None:
+            return constants
+
+        constants = {
+            name: torch.tensor(value, dtype=dtype, device=device)
+            for name, value in self._dcp_remap_scalar_values.items()
+        }
+        self._dcp_remap_scalar_tensors[key] = constants
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]: this should be seperated as another function

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 8]
```diff
@@ -578,3 +594,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+        if order_and_count is None:
+            order = torch.arange(topk_count, dtype=dtype, device=device)
+            count = torch.tensor(topk_count, dtype=dtype, device=device)
+            order_and_count = (order, count)
+            self._dcp_remap_order_buffers[key] = order_and_count
+        return order_and_count
+
+    def _map_indices_to_local(self, topk_indices: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
+        constants = self._get_dcp_remap_constants(topk_indices.device, torch.float32)
+        topk_float = topk_indices.to(torch.float32)
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]:
```suggestion
        topk_indices_fp32 = topk_indices.to(torch.float32)
```

File: vllm_ascend/attention/sfa_v1.py [hunk 9]
```diff
@@ -73,6 +75,14 @@
     "aclnn_input_offset",
 )
 
+DCPQueryGatherContext = tuple[
```
Comment (MengqingCao) [inline vllm_ascend/attention/sfa_v1.py]: plz add note to describe the semantics of these variables.

File: vllm_ascend/worker/model_runner_v1.py [hunk 10]
```diff
@@ -450,6 +451,21 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
             self.positions = torch.zeros(
                 max_buffer_num_tokens, dtype=torch.int64, device=self.device)
             
+        self.sfa_dcp_replicated_indexer_size = 1
+        sfa_dcp_replicated_indexer = self.use_sparse and (self.vllm_config.additional_config or {}).get(
+            "sfa_dcp_replicated_indexer", False
+        )
+        if sfa_dcp_replicated_indexer:
+            assert self.pcp_size == 1 and \
+            self.dcp_size == self.parallel_config.tensor_parallel_size
```
Comment (MengqingCao) [inline vllm_ascend/worker/model_runner_v1.py]: plz add assertion message for this

File: vllm_ascend/worker/model_runner_v1.py [hunk 11]
```diff
@@ -450,6 +451,21 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
...
+            if self.use_sparse_c8_indexer and get_ascend_device_type() == AscendDeviceType.A5:
+                raise NotImplementedError(
+                    "SFA DCP with sparse C8 LightningIndexer cache is not supported on A5 yet. "
+                    "A5 uses the fused CKV quant sparse attention path, which needs a separate DCP LSE merge."
+                )
```
Comment (MengqingCao) [inline vllm_ascend/worker/model_runner_v1.py]: we should add this check in the `AscendConfig`, instead of `modelrunner`

File: vllm_ascend/worker/model_runner_v1.py [hunk 12]
```diff
@@ -450,6 +451,21 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
             self.positions = torch.zeros(
                 max_buffer_num_tokens, dtype=torch.int64, device=self.device)
             
+        self.sfa_dcp_replicated_indexer_size = 1
+        sfa_dcp_replicated_indexer = self.use_sparse and (self.vllm_config.additional_config or {}).get(
+            "sfa_dcp_replicated_indexer", False
```
Comment (MengqingCao) [inline vllm_ascend/worker/model_runner_v1.py]: seems we could just use `enable_sfa_dcp_replicated_indexer`, we could add the check of sparse attention in that fuction.

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 13]
```diff
@@ -578,3 +594,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+
+    def _start_dcp_query_gather(
+        self,
+        ql_nope: torch.Tensor,
+        q_pe: torch.Tensor,
+        attn_metadata: M,
+    ) -> DCPQueryGatherContext:
+        assert attn_metadata.dcp_context is not None
+        query_gather_dim = 0 if self.enable_dsa_cp else 1
+        assert self.dcp_group is not None
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py]: plz add assertion message

File: vllm_ascend/attention/context_parallel/sfa_cp.py [hunk 14]
```diff
@@ -578,3 +594,672 @@ def _get_full_kv(self, k, attn_metadata: M):
...
+            )
+
+        ql_nope_dim = ql_nope.shape[-1]
+        q_pe_dim = q_pe.shape[-1]
+        # Avoid back-to-back DCP all_gather calls for the two SFA query
+        # fragments. On Ascend the separate gathers can leave SFA with an
+        # incomplete stream dependency on the first prefill. DSA-CP restores
+        # token shards on dim 0; native DCP restores query shards on dim 1.
+        fused_q = torch.cat([ql_nope, q_pe], dim=-1).contiguous()
+        gathered, handle, restore_perm = self._all_gather_dim_async(fused_q, query_gather_dim)
```
Comment (MengqingCao) [inline vllm_ascend/attention/context_parallel/sfa_cp.py:1122]: won't we need split the `ql_nope` and `q_pe` here?

**Review by maoxx241 — APPROVED** [review]
Confirmed the operator path supports this functionality. For this change, removing the validation is sufficient, so the current direction looks good to me.

Approved.

File: vllm_ascend/attention/context_parallel/sfa_cp.py (hunk 14 above)
Comment (MengqingCao) [reply vllm_ascend/attention/context_parallel/sfa_cp.py:1122]: ignore this plz, I find the split in `_finish_all_gather_query_for_dcp`

File: vllm_ascend/attention/context_parallel/sfa_cp.py (hunk 3 above)
Comment (MengqingCao) [reply vllm_ascend/attention/context_parallel/sfa_cp.py:743]: okay, it make sense then

**Review by MengqingCao — APPROVED** [review]
LGTM now, plz add the test case for this feature, thx!

**Comment by pisceskkk** [comment]
/e2e tests/e2e/pull_request/four_cards
[Bot]: e2e command triggered. See [workflow run](https://github.com/vllm-project/vllm-ascend/actions/runs/29023096622) for details.

**Review by MengqingCao — APPROVED** [review]

**Review by MengqingCao — APPROVED** [review]

**Comment by pisceskkk** [comment]
/cherry-pick releases/v0.23.0
[Bot]: cherry-pick command failed. Please check the [workflow run](https://github.com/vllm-project/vllm-ascend/actions/runs/29081461162) for details.
